### PR TITLE
Fix #31764: Show appropriate error message when fonts pubspec.yaml isn't iterable

### LINKED ISSUE #31764: Typo in pubspec.yaml fonts leads to TypeError instead of helpful error message.
I can't run my app on my device,when i click the run button, there was no gradle build info tips. and 
no error on my terminal,but only some info in my logs, floowing is the log 

Flutter crash report; please file at https://github.com/flutter/flutter/issues.

## command

flutter --no-color run --machine --track-widget-creation --device-id=GSL0216A23005048 lib\main.dart

## exception

_TypeError: type 'YamlMap' is not a subtype of type 'Iterable<dynamic>'

```
#0      _validateFonts (package:flutter_tools/src/flutter_manifest.dart:409:47)
#1      _validateFlutter (package:flutter_tools/src/flutter_manifest.dart:354:9)
#2      _validate (package:flutter_tools/src/flutter_manifest.dart:312:9)
<asynchronous suspension>
#3      FlutterManifest._createFromYaml (package:flutter_tools/src/flutter_manifest.dart:45:40)
<asynchronous suspension>
#4      FlutterManifest.createFromString (package:flutter_tools/src/flutter_manifest.dart:40:12)
<asynchronous suspension>
#5      FlutterManifest.createFromPath (package:flutter_tools/src/flutter_manifest.dart:34:12)
<asynchronous suspension>
#6      FlutterProject._readManifest (package:flutter_tools/src/project.dart:144:60)
<asynchronous suspension>
#7      FlutterProject.fromDirectory (package:flutter_tools/src/project.dart:44:44)
<asynchronous suspension>
#8      FlutterProject.current (package:flutter_tools/src/project.dart:55:46)
#9      FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:542:59)
<asynchronous suspension>
#10     FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:489:33)
<asynchronous suspension>
#11     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:142:29)
<asynchronous suspension>
#12     _rootRun (dart:async/zone.dart:1124:13)
#13     _CustomZone.run (dart:async/zone.dart:1021:19)
#14     _runZoned (dart:async/zone.dart:1516:10)
#15     runZoned (dart:async/zone.dart:1463:12)
#16     AppContext.run (package:flutter_tools/src/base/context.dart:141:18)
<asynchronous suspension>
#17     FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:480:20)
#18     CommandRunner.runCommand (package:args/command_runner.dart:196:27)
<asynchronous suspension>
#19     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:396:21)
<asynchronous suspension>
#20     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:142:29)
<asynchronous suspension>
#21     _rootRun (dart:async/zone.dart:1124:13)
#22     _CustomZone.run (dart:async/zone.dart:1021:19)
#23     _runZoned (dart:async/zone.dart:1516:10)
#24     runZoned (dart:async/zone.dart:1463:12)
#25     AppContext.run (package:flutter_tools/src/base/context.dart:141:18)
<asynchronous suspension>
#26     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:356:19)
<asynchronous suspension>
#27     CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:111:29)
#28     new Future.sync (dart:async/future.dart:224:31)
#29     CommandRunner.run (package:args/command_runner.dart:111:11)
#30     FlutterCommandRunner.run (package:flutter_tools/src/runner/flutter_command_runner.dart:242:18)
#31     run.<anonymous closure> (package:flutter_tools/runner.dart:60:20)
<asynchronous suspension>
#32     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:142:29)
<asynchronous suspension>
#33     _rootRun (dart:async/zone.dart:1124:13)
#34     _CustomZone.run (dart:async/zone.dart:1021:19)
#35     _runZoned (dart:async/zone.dart:1516:10)
#36     runZoned (dart:async/zone.dart:1463:12)
#37     AppContext.run (package:flutter_tools/src/base/context.dart:141:18)
<asynchronous suspension>
#38     runInContext (package:flutter_tools/src/context_runner.dart:50:24)
<asynchronous suspension>
#39     run (package:flutter_tools/runner.dart:51:10)
#40     main (package:flutter_tools/executable.dart:60:9)
<asynchronous suspension>
#41     main (file:///D:/FlutterSdk/flutter/packages/flutter_tools/bin/flutter_tools.dart:8:3)
#42     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:298:32)
#43     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:171:12)
```

## flutter doctor

```
[✓] Flutter (Channel beta, v1.4.9-hotfix.1, on Microsoft Windows [Version 6.1.7601], locale zh-CN)
    • Flutter version 1.4.9-hotfix.1 at D:\FlutterSdk\flutter
    • Framework revision 88fa7ea403 (2 weeks ago), 2019-04-11 14:01:46 -0700
    • Engine revision 4737fc5cd8
    • Dart version 2.2.1 (build 2.2.1-dev.4.1 None)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at C:\Users\dell\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: D:\AndroidStudio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[✓] Android Studio (version 3.2)
    • Android Studio at D:\AndroidStudio
    • Flutter plugin version 31.3.1
    • Dart plugin version 181.5656
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[✓] Connected device (1 available)
    • FRD AL10 • GSL0216A23005048 • android-arm64 • Android 8.0.0 (API 26)

• No issues found!
```

### PR DESCRIPTION
**Pull request by blasten**
Currently, the pubspec.yaml validator checks if "fonts" isn't a YamlList, but it doesn't exit the loop.  This causes the inner loop to run when `fontMap['fonts']` isn't a list, which results in this exception: `TypeError: type 'YamlMap' is not a subtype of type 'Iterable'`

Fixes #31764

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Please add a description to the PR briefly describing the fix for the issue

File: packages/flutter_tools/test/flutter_manifest_test.dart [hunk 1]
```diff
@@ -478,6 +481,28 @@ flutter:
...
+      const String manifest = '''
+name: test
+dependencies:
+  flutter:
+    sdk: flutter
+flutter:
+  fonts:
+    - family: foo
+      fonts:
+        -asset: a/bar
```
Comment (dnfield) [inline packages/flutter_tools/test/flutter_manifest_test.dart]: Should this be `- asset:`?

Comment (jonahwilliams) [reply packages/flutter_tools/test/flutter_manifest_test.dart]: This is the error case we're testing for from #31764

File: packages/flutter_tools/test/flutter_manifest_test.dart [hunk 2]
```diff
@@ -478,6 +481,26 @@ flutter:
...
+''';
+      final dynamic flutterManifest = FlutterManifest.createFromString(manifest);
+
+      expect(flutterManifest, null);
+      expect(logger.errorText, contains('Expected "fonts" to either be null or a list.'));
+
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/flutter_manifest_test.dart]:
```suggestion
```

File: packages/flutter_tools/test/flutter_manifest_test.dart [hunk 3]
```diff
@@ -478,6 +481,26 @@ flutter:
...
+dependencies:
+  flutter:
+    sdk: flutter
+flutter:
+  fonts:
+    - family: foo
+      fonts:
+        -asset: a/bar
+''';
+      final dynamic flutterManifest = FlutterManifest.createFromString(manifest);
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/flutter_manifest_test.dart]:
```suggestion
      final FlutterManifest flutterManifest = FlutterManifest.createFromString(manifest);
```

**Comment by blasten** [comment]
Done :)

File: packages/flutter_tools/lib/src/flutter_manifest.dart [hunk 4]
```diff
@@ -398,11 +398,13 @@ void _validateFonts(YamlList fonts, List<String> errors) {
     if (fontMap['family'] != null && fontMap['family'] is! String) {
       errors.add('Font family must either be null or a String.');
     }
-    if (fontMap['fonts'] != null && fontMap['fonts'] is! YamlList) {
-      errors.add('Expected "fonts" to either be null or a list.');
-    }
     if (fontMap['fonts'] == null) {
       continue;
+    } else {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/flutter_manifest.dart]: Combine the if and else statement here

File: packages/flutter_tools/test/flutter_manifest_test.dart [hunk 5]
```diff
@@ -478,6 +481,25 @@ flutter:
         expectedBuildNumber: null,
       );
     });
+
+    testUsingContext('returns proper error when font detail is malformed', () async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/flutter_manifest_test.dart]: Might add more context or a link to the original issue in a comment, since the error case is hard to parse out visually

File: packages/flutter_tools/test/flutter_manifest_test.dart (hunk 2 above)
Comment (blasten) [reply packages/flutter_tools/test/flutter_manifest_test.dart]: Done.

File: packages/flutter_tools/test/flutter_manifest_test.dart (hunk 3 above)
Comment (blasten) [reply packages/flutter_tools/test/flutter_manifest_test.dart]: Done.

File: packages/flutter_tools/lib/src/flutter_manifest.dart (hunk 4 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: Done.

File: packages/flutter_tools/test/flutter_manifest_test.dart (hunk 5 above)
Comment (blasten) [reply packages/flutter_tools/test/flutter_manifest_test.dart]: Great point. Thanks, Done.

**Review by jonahwilliams — APPROVED** [review]
LGTM
